frontend/catalyst/cuda/primitives/__init__.py

```diff
@@ -0,0 +1,578 @@
+# SKIP
+
+# Just-In-Time (JIT) compile `self` (`Kernel`) and call the kernel function at the provided
+# concrete arguments.
+# __call__(self: cudaq.Kernel, *args) -> None
+# SKIP
+
+# Apply a x gate to the given target qubit or qubits
+# x(self: cudaq.Kernel, target: cudaq.QuakeValue) -> None
+
+
+def make_primitive_for_gate():
+    """Just a function that wraps the functionality of making the kernel_inst primitive.
+
+    This function will return:
+      * gate_func: A convenience function for binding
+      * kernel_gate_p: A JAX primitive for quantum gates.
+    """
+    kernel_gate_p = jax.core.Primitive("kernel_inst")
+    kernel_gate_p.multiple_results = True
+
+    def gate_func(kernel, *qubits_or_params, inst=None, qubits_len=-1):
+        """Convenience.
+
+        Quantum operations in CUDA-quantum return no values. But JAXPR expects return values.
+        We can just say that multiple_results = True and return an empty tuple.
+        """
+        kernel_gate_p.bind(kernel, *qubits_or_params, inst=inst, qubits_len=qubits_len)
+        return tuple()
+
+    @kernel_gate_p.def_impl
+    def gate_impl(kernel, *qubits_or_params, inst=None, qubits_len=-1):
+        """Concrete implementation."""
+        assert inst and qubits_len > 0
+        method = getattr(cudaq.Kernel, inst)
+        targets = qubits_or_params[:qubits_len]
+        params = qubits_or_params[qubits_len:]
+        method(kernel, *params, *targets)
+        return tuple()
+
+    @kernel_gate_p.def_abstract_eval
+    def gate_abs(_kernel, *_qubits_or_params, inst=None, qubits_len=-1):
+        """Abstract evaluation."""
+        return tuple()
+
+    return gate_func, kernel_gate_p
+
+
+cuda_inst, cuda_inst_p = make_primitive_for_gate()
+
+
+def make_primitive_for_m(gate: str):
+    """A single function to make primitives for all measurement basis.
+
+    Args:
+      * gate (str): Either "x", "y", or "z" are valid values.
+    Returns:
+      * A function that binds a primitive.
+      * the primitive itself.
+    """
+
+    assert gate in {"x", "y", "z"}
+    gate = f"m{gate}"
+    kernel_gate_p = jax.core.Primitive(f"kernel_{gate}")
+    method = getattr(cudaq.Kernel, gate)
+
+    def gate_func(kernel, target):
+        """Convenience."""
+        return kernel_gate_p.bind(kernel, target)
+
+    @kernel_gate_p.def_impl
+    def gate_impl(kernel, target):
+        """Concrete implementation."""
+        return method(kernel, target)
+
+    @kernel_gate_p.def_abstract_eval
+    def gate_abs(_kernel, _target):
+        """Abstract evaluation."""
+        return AbsCudaValue()
+
+    return gate_func, kernel_gate_p
+
+
+mx_call, mx_p = make_primitive_for_m("x")
+my_call, my_p = make_primitive_for_m("y")
+mz_call, mz_p = make_primitive_for_m("z")
+
+
+cudaq_sample_p = jax.core.Primitive("cudaq_sample")
+cudaq_counts_p = jax.core.Primitive("cudaq_counts")
+cudaq_counts_p.multiple_results = True
```

Comment:
do you think it's better to move them all to the end of the file?

Reply:
I like to keep definitions close to their use.